src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -9842,253 +10110,327 @@ public static Vector512<T> Invoke(Vector512<T> x, Vector512<T> y)
+            sizeof(T) == sizeof(int) ? Vector128.LessThan(indices1.AsInt32(), indices2.AsInt32()).As<int, T>() :
+            sizeof(T) == sizeof(short) ? Vector128.LessThan(indices1.AsInt16(), indices2.AsInt16()).As<short, T>() :
+            sizeof(T) == sizeof(byte) ? Vector128.LessThan(indices1.AsByte(), indices2.AsByte()).As<byte, T>() :
+            throw new NotSupportedException();
 
-                maxIndex = ElementWiseSelect(greaterThanMask.AsInt32(), maxIndex, curIndex);
-            }
+        private static Vector256<T> IndexLessThan<T>(Vector256<T> indices1, Vector256<T> indices2) =>
+            sizeof(T) == sizeof(long) ? Vector256.LessThan(indices1.AsInt64(), indices2.AsInt64()).As<long, T>() :
+            sizeof(T) == sizeof(int) ? Vector256.LessThan(indices1.AsInt32(), indices2.AsInt32()).As<int, T>() :
+            sizeof(T) == sizeof(short) ? Vector256.LessThan(indices1.AsInt16(), indices2.AsInt16()).As<short, T>() :
+            sizeof(T) == sizeof(byte) ? Vector256.LessThan(indices1.AsByte(), indices2.AsByte()).As<byte, T>() :
+            throw new NotSupportedException();
 
+        private static Vector512<T> IndexLessThan<T>(Vector512<T> indices1, Vector512<T> indices2) =>
+            sizeof(T) == sizeof(long) ? Vector512.LessThan(indices1.AsInt64(), indices2.AsInt64()).As<long, T>() :
+            sizeof(T) == sizeof(int) ? Vector512.LessThan(indices1.AsInt32(), indices2.AsInt32()).As<int, T>() :
+            sizeof(T) == sizeof(short) ? Vector512.LessThan(indices1.AsInt16(), indices2.AsInt16()).As<short, T>() :
+            sizeof(T) == sizeof(byte) ? Vector512.LessThan(indices1.AsByte(), indices2.AsByte()).As<byte, T>() :
+            throw new NotSupportedException();
+
+        /// <summary>Returns the index of MathF.Max(x, y)</summary>
+        internal readonly struct IndexOfMaxOperator<T> : IIndexOfOperator<T> where T : INumber<T>
+        {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            public static int Invoke(ref float result, float current, int resultIndex, int curIndex)
+            public static void Invoke(ref Vector128<T> result, Vector128<T> current, ref Vector128<T> resultIndex, Vector128<T> curIndex)
             {
-                if (result == current)
+                Vector128<T> greaterThanMask = Vector128.GreaterThan(result, current);
+                Vector128<T> equalMask = Vector128.Equals(result, current);
+
+                if (equalMask != Vector128<T>.Zero)
                 {
-                    if (IsNegative(result) && !IsNegative(current))
+                    Vector128<T> lessThanIndexMask = IndexLessThan(resultIndex, curIndex);
+                    if (typeof(T) == typeof(float) || typeof(T) == typeof(double))
                     {
-                        result = current;
-                        return curIndex;
+                        greaterThanMask |= (equalMask & (~IsNegative(result) | IsNegative(current)) & lessThanIndexMask);
```

Comment:
I'm not sure if we pattern recognize `~Compare` for vectors today. It might be better to make it an explicit `IsPositive(result)` check instead

Reply:
I think having a comment explaining what this is doing for future readers would be beneficial and would reduce the cognitive strain of understanding the truth table we're building...

The general premise is that `+0 == -0` but for the purposes of `Min` and `Max` we need to treat `+0 > -0`. Additionally, for any inputs where the values are equal (factoring in the required `+0 > -0` rule), we need to pick the lesser index.

Thus we need to do effectively:
```csharp
// At this point:
// * greaterThanMask means x > y for everything except 0
// * equalMask means that x == y including for 0

// Build a mask that is true if the signs differ and false if they are the same
// 0 ^ 0    0    both are positive  
// 0 ^ 1    1    current is negative
// 1 ^ 0    1    result is negative
// 1 ^ 1    0    both are negative
Vector128<T> resultNegativeMask = IsNegative(result);
Vector128<T> signsDifferMask = resultNegativeMask  ^ IsNegative(current);

// Build an equality mask that excludes where the signs differed (excludes +0 == -0)
// 0 & ~0    0 & 1    0    not equal
// 0 & ~1    0 & 0    0    not equal
// 1 & ~0    1 & 1    1    equal and signs are the same
// 1 & ~1    1 & 0    0    equal and signs differed
Vector128<T> actualEqualMask = Vector128.AndNot(equalMask, signsDifferMask);

// Include the positive cases where they were equal but the signs differed (include +0 > -0)
// 0 & (0 & ~0)    0 & (0 & 1)    0 & 0    0    not equal (greater than mask already correct)
// 0 & (0 & ~1)    0 & (0 & 0)    0 & 0    0    not equal (greater than mask already correct)
// 0 & (1 & ~0)    0 & (1 & 1)    0 & 1    0    not equal (greater than mask already correct)
// 0 & (1 & ~1)    0 & (1 & 0)    0 & 0    0    not equal (greater than mask already correct)
// 1 & (0 & ~0)    1 & (0 & 1)    1 & 0    0    equal and signs are the same (greater than mask already correct)
// 1 & (0 & ~1)    1 & (0 & 0)    1 & 0    0    equal and signs are the same (greater than mask already correct)
// 1 & (1 & ~0)    1 & (1 & 1)    1 & 1    1    equal, signs differ, result is positive (set bits so we result is selected)
// 1 & (1 & ~1)    1 & (1 & 0)    1 & 0    0    equal, signs differ, result is negative (keep bits clear so current is selected)
greaterThanMask |=equalMask & Vector128.AndNot(signsDifferMask, resultNegativeMask);

// For cases that were actually equal, pick the one with the lesser index
greaterThanMask |= (actualEqualMask & lessThanIndexMask);
```

Reply:
(I don't think the full truth table is necessary, it just helped when I was working through the logic myself)